scripts/train_tokenizer.py

```diff
@@ -0,0 +1,131 @@
+import argparse
+import os
+
+from datasets import load_dataset
+from tqdm.auto import tqdm
+from transformers import LlamaTokenizerFast
+from tokenizers import SentencePieceBPETokenizer
+from tqdm import trange
+
+from delphi.train.tokenizer import train_vocab, get_tokenizer_model_path
+
+
+def main(
+    vocab_size: int,
+    dataset_name: str,
+    train_size: float,
+    username: str,
+    token: str,
+    funct_test: bool = False,
+):
+    """
+    Trains a SentencePiece tokenizer on a dataset.
+    And uploads the resulting tokenizer to huggingface.
+    Args:
+    - vocab_size: The vocabulary size of the resulting tokenizer
+    - dataset_name: The name of the dataset from which validation set will be loaded
+    - train_size: The amount of the dataset that should be used for training
+    - username: Hugging Face API username
+    - token: Hugging Face API token
+    """
+    train_ds = load_dataset(dataset_name)["train"]
+    if train_size < 1.0:
+        train_ds = train_ds.train_test_split(train_size=train_size)["train"]
+        
+    tokenizer_model_path = get_tokenizer_model_path(
+        vocab_size=vocab_size,
+    )
+    if not os.path.isfile(tokenizer_model_path):
+        train_vocab(
+            vocab_size=vocab_size,
+            dataset=train_ds,
+        )
+        
+    import sentencepiece as spm
+    tokenizer_model_path = get_tokenizer_model_path(vocab_size=vocab_size)
+    sp_model = spm.SentencePieceProcessor(model_file=tokenizer_model_path)
+    
+    # export 'vocab' and 'merges'
+    vocab = {sp_model.id_to_piece(index): index for index in trange(sp_model.GetPieceSize())}
+    merges = []
+    for piece_l in tqdm(vocab.keys(), total=sp_model.GetPieceSize()):
+        for piece_r in vocab.keys():
+            merge = f"{piece_l}{piece_r}"
+            piece_id = vocab.get(merge, None)
+            if piece_id:
+                merges += [(piece_l, piece_r, piece_id)]
+    merges = sorted(merges, key=lambda val: val[2])
+    merges = [(val[0], val[1]) for val in merges]
```

Comment:
any idea what is this doing?

Reply:
This generates the components HF needs to initialise the SentencePieceBPETokenizer. `vocab` are all tokens, and `merges` are all combinations of tokens that exist as another token in the vocabulary (e.g. "hel" + "lo" = "hello")